atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/expectations/TrackingKeyValueServiceImpl.java

```diff
@@ -0,0 +1,219 @@
+import com.palantir.atlasdb.keyvalue.impl.ForwardingKeyValueService;
+import com.palantir.atlasdb.transaction.api.expectations.TransactionReadInfo;
+import com.palantir.common.base.ClosableIterator;
+import com.palantir.common.exception.AtlasDbDependencyException;
+import com.palantir.util.paging.TokenBackedBasicResultsPage;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+
+public class TrackingKeyValueServiceImpl extends ForwardingKeyValueService implements TrackingKeyValueService {
+    KeyValueService delegate;
+    KeyValueServiceDataTracker tracker = new KeyValueServiceDataTracker();
+
+    public TrackingKeyValueServiceImpl(KeyValueService delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public KeyValueService delegate() {
+        return delegate;
+    }
+
+    @Override
+    public TransactionReadInfo getOverallReadInfo() {
+        return tracker.getReadInfo();
+    }
+
+    @Override
+    public ImmutableMap<TableReference, TransactionReadInfo> getReadInfoByTable() {
+        return tracker.getReadInfoByTable();
+    }
+
+    @Override
+    public ListenableFuture<Map<Cell, Value>> getAsync(TableReference tableRef, Map<Cell, Long> timestampByCell) {
+        return Futures.transform(
+                delegate.getAsync(tableRef, timestampByCell),
+                valueByCell -> {
+                    tracker.readForTable(tableRef, "getAsync", ExpectationsMeasuringUtils.sizeInBytes(valueByCell));
+                    return valueByCell;
+                },
+                MoreExecutors.directExecutor());
+    }
+
+    @Override
+    public Map<Cell, Value> getRows(
+            TableReference tableRef, Iterable<byte[]> rows, ColumnSelection columnSelection, long timestamp) {
+        Map<Cell, Value> result = delegate.getRows(tableRef, rows, columnSelection, timestamp);
+        tracker.readForTable(tableRef, "getRows", ExpectationsMeasuringUtils.sizeInBytes(result));
+        return result;
+    }
+
+    @Override
+    public Map<byte[], RowColumnRangeIterator> getRowsColumnRange(
+            TableReference tableRef,
+            Iterable<byte[]> rows,
+            BatchColumnRangeSelection batchColumnRangeSelection,
+            long timestamp) {
+        tracker.callForTable(tableRef);
+        Map<byte[], RowColumnRangeIterator> result =
+                delegate.getRowsColumnRange(tableRef, rows, batchColumnRangeSelection, timestamp);
+        // todo(aalouane) this throws if delegate returns an immutable map or a map that does not implement replaceAll
+        // not sure what to do here
+        result.replaceAll((rowsRead, iterator) -> {
+            tracker.partialReadForTable(tableRef, rowsRead.length);
+            return new TrackingRowColumnRangeIterator(
+                    iterator, partialReadForTableConsumer(tableRef), ExpectationsMeasuringUtils::sizeInBytes);
+        });
+        return result;
```

Comment:
The following is relevant to the comment above, but also applies to this class

As discussed previously, we might need to wrap all tracking calls in try/finally